certbot/cert_manager.py

```diff
@@ -139,6 +150,95 @@ def update_certs_for_domain_matches(candidate_lineage, rv):
 
     return _search_lineages(config, update_certs_for_domain_matches, (None, None))
 
+def match_and_check_overlaps(cli_config, acceptable_matches, match_func, rv_func):
+    """ Searches through all lineages for a match, and checks for duplicates.
+    If a duplicate is found, an error is raised, as performing operations on lineages
+    that have their properties incorrectly duplicated elsewhere is probably a bad idea.
+
+    :param list acceptable_matches: a list of functions that specify acceptable matches
+    :param function match_func: specifies what to match
+    :param function rv_func: specifies what to return
+    """
+    def find_matches(candidate_lineage, return_value, acceptable_matches):
+        """Returns a list of matches using _search_lineages."""
+        acceptable_matches = [func(candidate_lineage) for func in acceptable_matches]
+        match = match_func(candidate_lineage)
+        if match in acceptable_matches:
+            return_value.append(rv_func(candidate_lineage))
+        return return_value
+
+    matched = _search_lineages(cli_config, find_matches, [], acceptable_matches)
+    if not matched:
+        raise errors.Error("No match found for cert-path {0}!".format(cli_config.cert_path))
+    elif len(matched) > 1:
+        raise errors.OverlappingMatchFound()
+    else:
+        return matched
+
+def cert_path_to_lineage(cli_config):
+    """ If config.cert_path is defined, find an appropriate value for config.certname
+    by searching through available files in config.renewal_configs_dir, and finding
+    one with an appropriate value for 'fullchain'."""
+
+    def archive_files(candidate_lineage, filetype):
+        """ In order to match things like:
+            /etc/letsencrypt/archive/example.com/chain1.pem"""
+        archive_dir = os.path.join(candidate_lineage.archive_dir,
+                os.path.basename(candidate_lineage.cert_path[0]))
+        pattern = [os.path.join(archive_dir, f) for f in os.listdir(archive_dir)
```

Comment:
This code should maybe be more robustly tested, because `archive_dir` is a file here, so `os.listdir` is unlikely to be getting what it's intended to be getting.

Also, since it's using implementation details of how storage works, it should either be in `storage.py` or have extra-robust tests so that if someone went in and refactored storage, the tests for this would catch that and fail.